log-generator/generator.py

```diff
@@ -228,12 +228,17 @@ def generate_batch(users, items):
 
     prev_hour = datetime.now().replace(microsecond=0, second=0, minute=0) - timedelta(hours=1)
     logs = generate_logs(users, prev_hour, lambda *a: line_generator_item_access(*a, items=items))
-    logs_as_csv_string = "\n".join((",".join(line) for line in logs))
 
-    s3 = boto3.client("s3")
-
-    with io.BytesIO(logs_as_csv_string.encode('utf-8')) as f:
-        s3.upload_fileobj(f, S3_BUCKET, S3_PREFIX + prev_hour.strftime("/%Y/%m/%d/%H.csv"))
+    with open("batch.csv", "w") as batch:
```

Comment:
@modelflat 

- `csv.DictWriter()` expects a text file handle;
- `boto3.client("s3").upload_fileobj()` expects a binary file handle.

So, we'd need to convert `io.StringIO` to `io.BytesIO`:

```py
with io.StringIO() as batch:
    logs = map(json.loads, logs)
    first, rest = next(logs), logs
    writer = csv.DictWriter(batch, fieldnames=first.keys())
    writer.writeheader()
    writer.writerow(first)
    writer.writerows(rest)
    boto3.client("s3").upload_fileobj(
        io.BytesIO(batch.getvalue().encode()),
        S3_BUCKET,
        S3_PREFIX + prev_hour.strftime("/%Y/%m/%d/%H.csv")
    )
```

I figured it was better to go for a plain file writing instead.

Reply:
I understand that this is a toy project, so it does not really matter but I would still prefer avoid creating an actual file in favor of using in-memory streams (especially since this does not imply writing much more code, and it is more correct IMO). What if you decide to invoke this program multiple times in parallel?

Reply:
@modelflat Done!